README.md
# Terraform "filesystem" provider

The *filesystem* Terraform provider manages local files and directories.

⚠️ This provider has been developed for educational purposes only, **it is not suitable for production use**.

## Installation

Note: the [Go](https://golang.org/) language compiler is required for building the provider.

Clone the repository into your `$GOPATH`:

```
$ mkdir -p $GOPATH/src/github.com/d2si-oss
$ git clone https://github.com/d2si-oss/terraform-provider-filesystem $GOPATH/src/github.com/d2si-oss/terraform-provider-filesystem
```

Enter the provider directory and build the provider

```
$ cd $GOPATH/src/github.com/d2si-oss/terraform-provider-filesystem
$ make build
```

Initialize the stack using the built provider

```
$ cd /path/to/terraform/stack
$ terraform init -plugin-dir=$GOPATH/bin
$ terraform plan
```

You should be all set!

## Configuration

### Resource "directory"

* `path` (required – type string): Path to the directory to be created
* `user` (optional – type string, default to current user): Directory owner user name
* `group` (optional – type string, default to current primary group): Directory owner group name
* `mode` (optional – type string, default `"0755"`): Permissions to apply to directory (in octal representation, e.g. 0755)
* `create_parents` (optional – type bool, default `false`): Create parent directories as needed

### Resource "file"

* `path` (required – type string): Path to the file to be created
* `user` (optional – type string, default to current user): File owner user name
* `group` (optional – type string, default to current primary group): File owner group name
* `mode` (optional – type string, default `"0644"`): Permissions to apply to file (in octal representation, e.g. 0644)
* `content` (optional – type string, default `""`): File content

## Example Usage

Using the following Terraform configuration:

```
$ cat test.tf
provider "filesystem" {
  # debug = false
}

resource "filesystem_directory" "test" {
  path = "/tmp/test/dir"
  user = "marc"
  group = "admin"
  mode = "0750"
  create_parents = true
}

resource "filesystem_file" "test" {
  path = "${filesystem_directory.test.path}/file"
  user = "marc"
  group = "admin"
  mode = "0640"
  content = <<EOF

                              /\__/\
                             /`    '\
                           === 0  0 ===
                             \  --  /
                            /        \
                           /          \
                           |           |
                           \  ||  ||  /
                            \_oo__oo_/#######o

EOF
}
```

Let's start by initializing the provider:

```
 $ terraform init

Initializing provider plugins...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

$ terraform apply -auto-approve
filesystem_directory.test: Creating...
  create_parents: "" => "true"
  group:          "" => "admin"
  mode:           "" => "020000000750"
  path:           "" => "/tmp/test/dir"
  user:           "" => "marc"
filesystem_directory.test: Creation complete after 0s (ID: f5aee067c62015c4883439792bf8d42139f0ee72f2802ee0d55af7bb6a676f21)
filesystem_file.test: Creating...
  content: "" => "491b3d14819e554aaa2c950d459c3e55b99690c2b132d9c681722135458416cf"
  group:   "" => "admin"
  mode:    "" => "0640"
  path:    "" => "/tmp/test/dir/file"
  user:    "" => "marc"
filesystem_file.test: Creation complete after 0s (ID: ea77e763acccecfd63e89cef535a799dfee4f9a67e606f512a03f568b5e419d2)

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```

Check that resources have been created according to our specifications:

```
$ tree /tmp/test/
/tmp/test/
└── dir
    └── file

1 directory, 1 file

$ stat /tmp/test/dir/ /tmp/test/dir/file
  File: /tmp/test/dir/
  Size: 96        	Blocks: 0          IO Block: 4194304 directory
Device: 1000004h/16777220d	Inode: 7489559     Links: 3
Access: (0750/drwxr-x---)  Uid: (  501/    marc)   Gid: (   80/   admin)
Access: 2018-02-08 15:19:16.494977804 +0100
Modify: 2018-02-08 15:19:03.154353901 +0100
Change: 2018-02-08 15:19:03.154353901 +0100
 Birth: 2018-02-08 15:19:03.148982056 +0100
  File: /tmp/test/dir/file
  Size: 362       	Blocks: 8          IO Block: 4194304 regular file
Device: 1000004h/16777220d	Inode: 7489560     Links: 1
Access: (0640/-rw-r-----)  Uid: (  501/    marc)   Gid: (   80/   admin)
Access: 2018-02-08 15:19:03.154332797 +0100
Modify: 2018-02-08 15:19:03.154427021 +0100
Change: 2018-02-08 15:19:03.154489856 +0100
 Birth: 2018-02-08 15:19:03.154332797 +0100

$ cat /tmp/test/dir/file

                              /\__/\
                             /`    '\
                           === 0  0 ===
                             \  --  /
                            /        \
                           /          \
                           |           |
                           \  ||  ||  /
                            \_oo__oo_/#######o

```

*"Introduce a little anarchy..."*

```
$ echo meow > /tmp/test/dir/file
$ cat /tmp/test/dir/file
meow
```

We now restore our file to the desired state:

```
$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

filesystem_directory.test: Refreshing state... (ID: f5aee067c62015c4883439792bf8d42139f0ee72f2802ee0d55af7bb6a676f21)
filesystem_file.test: Refreshing state... (ID: ea77e763acccecfd63e89cef535a799dfee4f9a67e606f512a03f568b5e419d2)

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  ~ filesystem_file.test
      content: "b0f0d8ff8cc965a7b70b07e0c6b4c028f132597196ae9c70c620cb9e41344106" => "491b3d14819e554aaa2c950d459c3e55b99690c2b132d9c681722135458416cf"


Plan: 0 to add, 1 to change, 0 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.

$ terraform apply -auto-approve
filesystem_directory.test: Refreshing state... (ID: f5aee067c62015c4883439792bf8d42139f0ee72f2802ee0d55af7bb6a676f21)
filesystem_file.test: Refreshing state... (ID: ea77e763acccecfd63e89cef535a799dfee4f9a67e606f512a03f568b5e419d2)
filesystem_file.test: Modifying... (ID: ea77e763acccecfd63e89cef535a799dfee4f9a67e606f512a03f568b5e419d2)
  content: "b0f0d8ff8cc965a7b70b07e0c6b4c028f132597196ae9c70c620cb9e41344106" => "491b3d14819e554aaa2c950d459c3e55b99690c2b132d9c681722135458416cf"
filesystem_file.test: Modifications complete after 0s (ID: ea77e763acccecfd63e89cef535a799dfee4f9a67e606f512a03f568b5e419d2)

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.

$ cat /tmp/test/dir/file

                              /\__/\
                             /`    '\
                           === 0  0 ===
                             \  --  /
                            /        \
                           /          \
                           |           |
                           \  ||  ||  /
                            \_oo__oo_/#######o

```

We're done, let's clean up:

```
$ terraform destroy -force
filesystem_directory.test: Refreshing state... (ID: f5aee067c62015c4883439792bf8d42139f0ee72f2802ee0d55af7bb6a676f21)
filesystem_file.test: Refreshing state... (ID: ea77e763acccecfd63e89cef535a799dfee4f9a67e606f512a03f568b5e419d2)
filesystem_file.test: Destroying... (ID: ea77e763acccecfd63e89cef535a799dfee4f9a67e606f512a03f568b5e419d2)
filesystem_file.test: Destruction complete after 0s
filesystem_directory.test: Destroying... (ID: f5aee067c62015c4883439792bf8d42139f0ee72f2802ee0d55af7bb6a676f21)
filesystem_directory.test: Destruction complete after 0s

Destroy complete! Resources: 2 destroyed.

$ ls /tmp/test/
$
```


CHANGELOG.md
## 0.1.0 (Feb 23, 2018)

* First release


filesystem/provider.go
package filesystem

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"

	"github.com/facette/logger"
	"github.com/hashicorp/terraform/helper/schema"
	"github.com/hashicorp/terraform/terraform"
)

type filesystemProvider struct {
	log *logger.Logger
}

var providerLogFile = "terraform-provider-filesystem.log"

func Provider() terraform.ResourceProvider {
	return &schema.Provider{
		Schema: map[string]*schema.Schema{
			"debug": {
				Type:        schema.TypeBool,
				Description: fmt.Sprintf("Enable provider debug logging (logs to file %s)", providerLogFile),
				Optional:    true,
				Default:     false,
			},
		},

		ResourcesMap: map[string]*schema.Resource{
			"filesystem_directory": resourceDirectory(),
			"filesystem_file":      resourceFile(),
		},

		ConfigureFunc: config,
	}
}

func config(d *schema.ResourceData) (interface{}, error) {
	var (
		p            filesystemProvider
		loggerConfig logger.FileConfig
		err          error
	)

	if d.Get("debug").(bool) {
		loggerConfig = logger.FileConfig{
			Level: "debug",
			Path:  providerLogFile,
		}
	}

	if p.log, err = logger.NewLogger(loggerConfig); err != nil {
		return nil, fmt.Errorf("unable to init provider debug logger: %s", err)
	}

	return p, nil
}

func hash(s string) string {
	sha := sha256.Sum256([]byte(s))
	return hex.EncodeToString(sha[:])
}


filesystem/resource_filesystem_file.go
package filesystem

import (
	"fmt"
	"io/ioutil"
	"os"
	"os/user"
	"strconv"
	"syscall"

	"github.com/hashicorp/terraform/helper/schema"
)

func resourceFile() *schema.Resource {
	return &schema.Resource{
		Schema: map[string]*schema.Schema{
			"path": {
				Type:        schema.TypeString,
				Description: "Path to the file to be created",
				Required:    true,
				ForceNew:    true,
			},
			"user": {
				Type:        schema.TypeString,
				Description: "File owner user name (default: current user)",
				Optional:    true,
				ForceNew:    false,
				DefaultFunc: func() (interface{}, error) {
					currentUser, err := user.Current()
					if err != nil {
						return nil, fmt.Errorf("unable to lookup current user name: %s", err)
					}
					return currentUser.Username, nil
				},
			},
			"group": {
				Type:        schema.TypeString,
				Description: "File owner group name",
				Optional:    true,
				ForceNew:    false,
				DefaultFunc: func() (interface{}, error) {
					currentUser, err := user.Current()
					if err != nil {
						return nil, fmt.Errorf("unable to lookup current user name: %s", err)
					}

					currentGroup, err := user.LookupGroupId(currentUser.Gid)
					if err != nil {
						return nil, fmt.Errorf("unable to lookup current user group name: %s", err)
					}
					return currentGroup.Name, nil
				},
			},
			"mode": {
				Type:        schema.TypeString,
				Description: "Permissions to apply to file (in octal representation, e.g. 0644)",
				Optional:    true,
				Default:     "0644",
				ForceNew:    false,
				ValidateFunc: func(i interface{}, k string) (ws []string, errors []error) {
					if _, err := strconv.ParseUint(i.(string), 8, 32); err != nil {
						errors = append(errors, fmt.Errorf("%q: invalid value", k))
					}
					return
				},
			},
			"content": {
				Type:        schema.TypeString,
				Description: "File content",
				Optional:    true,
				Default:     "",
				ForceNew:    false,
				StateFunc: func(v interface{}) string {
					return hash(v.(string))
				},
			},
		},

		Create: resourceFilesystemFileCreate,
		Read:   resourceFilesystemFileRead,
		Update: resourceFilesystemFileUpdate,
		Delete: resourceFilesystemFileDelete,
	}
}

func resourceFilesystemFileCreate(d *schema.ResourceData, meta interface{}) error {
	p := meta.(filesystemProvider)

	p.log.Debug("calling resourceFilesystemFileCreate()")

	fileMode, _ := strconv.ParseUint(d.Get("mode").(string), 8, 32)
	d.Set("mode", fmt.Sprintf("%#o", os.FileMode(fileMode)))

	file, err := os.OpenFile(d.Get("path").(string), os.O_RDWR|os.O_CREATE, os.FileMode(fileMode))
	if err != nil {
		return err
	}
	defer file.Close()

	if _, err := file.WriteString(d.Get("content").(string)); err != nil {
		return err
	}

	u, err := user.Lookup(d.Get("user").(string))
	if err != nil {
		return fmt.Errorf("unable to lookup file owner user information: %s", err)
	}
	uid, _ := strconv.Atoi(u.Uid)

	g, err := user.LookupGroup(d.Get("group").(string))
	if err != nil {
		return fmt.Errorf("unable to lookup file owner group information: %s", err)
	}
	gid, _ := strconv.Atoi(g.Gid)

	if err := file.Chown(uid, gid); err != nil {
		return fmt.Errorf("unable to change file user/group: %s", err)
	}

	d.SetId(hash(file.Name()))

	return nil
}

func resourceFilesystemFileRead(d *schema.ResourceData, meta interface{}) error {
	p := meta.(filesystemProvider)

	p.log.Debug("calling resourceFilesystemFileRead()")

	fileInfo, err := os.Stat(d.Get("path").(string))
	if err != nil {
		if os.IsNotExist(err) {
			d.SetId("")
			return nil
		}

		return err
	}
	d.Set("mode", fmt.Sprintf("%#o", fileInfo.Mode()))

	fileContent, err := ioutil.ReadFile(d.Get("path").(string))
	if err != nil {
		return err
	}
	d.Set("content", hash(string(fileContent)))

	u, err := user.LookupId(fmt.Sprintf("%d", fileInfo.Sys().(*syscall.Stat_t).Uid))
	if err != nil {
		return fmt.Errorf("unable to lookup file owner user information: %s", err)
	}
	d.Set("user", u.Username)

	g, err := user.LookupGroupId(fmt.Sprintf("%d", fileInfo.Sys().(*syscall.Stat_t).Gid))
	if err != nil {
		return fmt.Errorf("unable to lookup file owner group information: %s", err)
	}
	d.Set("group", g.Name)

	return nil
}

func resourceFilesystemFileUpdate(d *schema.ResourceData, meta interface{}) error {
	p := meta.(filesystemProvider)

	p.log.Debug("calling resourceFilesystemFileUpdate()")

	file, err := os.OpenFile(d.Get("path").(string), os.O_RDWR, 0666)
	if err != nil {
		return err
	}
	defer file.Close()

	if d.HasChange("mode") {
		fileMode, _ := strconv.ParseUint(d.Get("mode").(string), 8, 32)

		if err := file.Chmod(os.FileMode(fileMode)); err != nil {
			return err
		}
	}

	if d.HasChange("user") || d.HasChange("group") {
		u, err := user.Lookup(d.Get("user").(string))
		if err != nil {
			return fmt.Errorf("unable to lookup file owner user information: %s", err)
		}
		uid, _ := strconv.Atoi(u.Uid)

		g, err := user.LookupGroup(d.Get("group").(string))
		if err != nil {
			return fmt.Errorf("unable to lookup file owner group information: %s", err)
		}
		gid, _ := strconv.Atoi(g.Gid)

		if err := file.Chown(uid, gid); err != nil {
			return fmt.Errorf("unable to change file user/group: %s", err)
		}
	}

	if d.HasChange("content") {
		if err := file.Truncate(0); err != nil {
			return err
		}

		if _, err := file.WriteString(d.Get("content").(string)); err != nil {
			return err
		}
	}

	return nil
}

func resourceFilesystemFileDelete(d *schema.ResourceData, meta interface{}) error {
	p := meta.(filesystemProvider)

	p.log.Debug("calling resourceFilesystemFileDelete()")

	return os.Remove(d.Get("path").(string))
}


filesystem/resource_filesystem_directory.go
package filesystem

import (
	"fmt"
	"os"
	"os/user"
	"strconv"
	"syscall"

	"github.com/hashicorp/terraform/helper/schema"
)

func resourceDirectory() *schema.Resource {
	return &schema.Resource{
		Schema: map[string]*schema.Schema{
			"path": {
				Type:        schema.TypeString,
				Description: "Path to the directory to be created",
				Required:    true,
				ForceNew:    true,
			},
			"user": {
				Type:        schema.TypeString,
				Description: "Directory owner user name (default: current user)",
				Optional:    true,
				ForceNew:    false,
				DefaultFunc: func() (interface{}, error) {
					currentUser, err := user.Current()
					if err != nil {
						return nil, fmt.Errorf("unable to lookup current user name: %s", err)
					}
					return currentUser.Username, nil
				},
			},
			"group": {
				Type:        schema.TypeString,
				Description: "Directory owner group name",
				Optional:    true,
				ForceNew:    false,
				DefaultFunc: func() (interface{}, error) {
					currentUser, err := user.Current()
					if err != nil {
						return nil, fmt.Errorf("unable to lookup current user name: %s", err)
					}

					currentGroup, err := user.LookupGroupId(currentUser.Gid)
					if err != nil {
						return nil, fmt.Errorf("unable to lookup current user group name: %s", err)
					}
					return currentGroup.Name, nil
				},
			},
			"mode": {
				Type:        schema.TypeString,
				Description: "Permissions to apply to directory (in octal representation, e.g. 0755)",
				Optional:    true,
				Default:     "0755",
				ForceNew:    false,
				ValidateFunc: func(i interface{}, k string) (ws []string, errors []error) {
					if _, err := strconv.ParseUint(i.(string), 8, 32); err != nil {
						errors = append(errors, fmt.Errorf("%q: invalid value", k))
					}
					return
				},
				StateFunc: func(v interface{}) string {
					// We serialize the permissions including 'directory mode' (e.g. `020000000755`) or else
					// the internal format will always be found different from the state format (`0755`)
					permBits, _ := strconv.ParseUint(v.(string), 8, 32)
					return fmt.Sprintf("%#o", os.ModeDir|os.FileMode(permBits))
				},
			},
			"create_parents": {
				Type:        schema.TypeBool,
				Description: "Create parent directories as needed",
				Optional:    true,
				Default:     false,
				ForceNew:    false,
			},
		},

		Create: resourceFilesystemDirectoryCreate,
		Read:   resourceFilesystemDirectoryRead,
		Update: resourceFilesystemDirectoryUpdate,
		Delete: resourceFilesystemDirectoryDelete,
	}
}

func resourceFilesystemDirectoryCreate(d *schema.ResourceData, meta interface{}) error {
	p := meta.(filesystemProvider)

	p.log.Debug("calling resourceFilesystemDirectoryCreate()")

	dirMode, _ := strconv.ParseUint(d.Get("mode").(string), 8, 32)

	if d.Get("create_parents").(bool) {
		if err := os.MkdirAll(d.Get("path").(string), os.FileMode(dirMode)); err != nil {
			return err
		}
	} else {
		if err := os.Mkdir(d.Get("path").(string), os.FileMode(dirMode)); err != nil {
			return err
		}
	}

	dir, err := os.OpenFile(d.Get("path").(string), os.O_RDONLY, os.FileMode(dirMode))
	if err != nil {
		return err
	}
	defer dir.Close()

	dirInfo, err := dir.Stat()
	if err != nil {
		return err
	}
	d.Set("mode", fmt.Sprintf("%#o", dirInfo.Mode()))

	u, err := user.Lookup(d.Get("user").(string))
	if err != nil {
		return fmt.Errorf("unable to lookup file owner user information: %s", err)
	}
	uid, _ := strconv.Atoi(u.Uid)

	g, err := user.LookupGroup(d.Get("group").(string))
	if err != nil {
		return fmt.Errorf("unable to lookup file owner group information: %s", err)
	}
	gid, _ := strconv.Atoi(g.Gid)

	if err := dir.Chown(uid, gid); err != nil {
		return fmt.Errorf("unable to change file user/group: %s", err)
	}

	d.SetId(hash(dir.Name()))

	return nil
}

func resourceFilesystemDirectoryRead(d *schema.ResourceData, meta interface{}) error {
	p := meta.(filesystemProvider)

	p.log.Debug("calling resourceFilesystemDirectoryRead()")

	dirInfo, err := os.Stat(d.Get("path").(string))
	if err != nil {
		if os.IsNotExist(err) {
			d.SetId("")
			return nil
		}

		return err
	}
	d.Set("mode", fmt.Sprintf("%#o", dirInfo.Mode()))

	u, err := user.LookupId(fmt.Sprintf("%d", dirInfo.Sys().(*syscall.Stat_t).Uid))
	if err != nil {
		return fmt.Errorf("unable to lookup directory owner user information: %s", err)
	}
	d.Set("user", u.Username)

	g, err := user.LookupGroupId(fmt.Sprintf("%d", dirInfo.Sys().(*syscall.Stat_t).Gid))
	if err != nil {
		return fmt.Errorf("unable to lookup directory owner group information: %s", err)
	}
	d.Set("group", g.Name)

	return nil
}

func resourceFilesystemDirectoryUpdate(d *schema.ResourceData, meta interface{}) error {
	p := meta.(filesystemProvider)

	p.log.Debug("calling resourceFilesystemDirectoryUpdate()")

	dir, err := os.OpenFile(d.Get("path").(string), os.O_RDONLY, 0666)
	if err != nil {
		return err
	}
	defer dir.Close()

	if d.HasChange("mode") {
		dirMode, _ := strconv.ParseUint(d.Get("mode").(string), 8, 32)

		if err := dir.Chmod(os.FileMode(dirMode)); err != nil {
			return err
		}
	}

	if d.HasChange("user") || d.HasChange("group") {
		u, err := user.Lookup(d.Get("user").(string))
		if err != nil {
			return fmt.Errorf("unable to lookup directory owner user information: %s", err)
		}
		uid, _ := strconv.Atoi(u.Uid)

		g, err := user.LookupGroup(d.Get("group").(string))
		if err != nil {
			return fmt.Errorf("unable to lookup directory owner group information: %s", err)
		}
		gid, _ := strconv.Atoi(g.Gid)

		if err := dir.Chown(uid, gid); err != nil {
			return fmt.Errorf("unable to change directory user/group: %s", err)
		}
	}

	return nil
}

func resourceFilesystemDirectoryDelete(d *schema.ResourceData, meta interface{}) error {
	p := meta.(filesystemProvider)

	p.log.Debug("calling resourceFilesystemDirectoryDelete()")

	return os.Remove(d.Get("path").(string))
}
